### DISCUSSION TITLE
Basic crafting

### DISCUSSION THREAD

**Discussion by Volte6** [GoMudEngine/GoMud#52] [Ideas]
# Goal

I'm interested in implementing basic crafting, ideally in a manner that can be more "automated" and require less hands on detailing on a per item basis.

If done correctly, it can be a useful skill for players to have to min/max or even just create useful items that can't be easily purchased, and at the same time be an in-game marketable skill, where one player creates a desirable object to sell to other players. 

_Note: When considering restrictions placed on "What can be crafted", its worth noting that the more restrictions applied, the less valuable and interesting the skill is, or the usefulness of salvaging in general._

# Mechanics

To achieve this I'm thinking there are two aspects to crafting:

* Salvaging Items - _Items should be broken down into components, with a degree of loss in the process. For example, a dagger may yield one "inorganic" component or otherwise. A fraction of components should be rendered into "junk" as a scrap rate._
* Crafting Items - _Crafting should require X number of core components, which will be combined to create an item with some potential of randomized statistics._

Crafting should be a time consuming process. While it's not fun to require a player to sit in a room for hours waiting for a result, it should be fine to require them to return at a later date to see the result.

# Components

This part is tricky. We don't want to specify all of the components on a per-item basis, either for crafting or for salvaging.

Since components may become plentiful, we probably want a way to group them into quantity bundles. Possibly use the `uses` attribute of items to keep track of quantity.

## Location restriction

We can specify at a "smithy" that certain weapon types or exact weapon id's can be crafted there, along with their cost in components. This is nice because it prevents crafting from getting out of control to where players can "craft anything". It's not the best option because it requires hand picking what can be crafted and spelling out what it will cost.

## Blueprints

We can create blueprints that function similar to a smithy, in that they specify the cost of a weapon. Players can buy/learn blueprints, thereby enabling the crafting of that object under the right conditions/expense.

## Skill based

Crafting could optionally be a per-item learned skill, where the process of salvaging an item lends itself to also learning how to construct that item with enough experience. This is dangerous, as players could learn to craft items that we don't want frequently "copied". We could restrict certain items but that's a lot of book keeping.

# Salvaging

Items should break down into components, with increasingly valuable or plentiful components being yielded by better objects that are salvaged. It may even be a worthwhile skill to train/develop to decrease the players scrap rate, or maximize the quality of the components.

The yield should be algorithmically generated. We can hopefully derive from the specifications of the item what kind of components to yield.

For example, a `weapon` with a `+2` damage modifier and applies `2 buffs` on crit should yield better components than a plain weapon such as a sharp stick. The component type would be `Inorganic`

Initially, I'm thinking keeping the quantity of yielded components identical across the board, such as `1 or 2` components (2 would be exceptional success at savaging, or if an item has a magical property, possibly yielding a second component of type `magical component`). The `QUALITY` of those components increases based on specs of the item being salvaged. 

## Component Types

* Organic Component: _Consumables such as food_
* Inorganic Component: _Used for crafting weapons, armor, and tools._
* Magical Component: _Used to imbue magical properties_

As a simple example, an `Organic Component` and a `Magical Component` are required to craft a `Red Potion`

## Component Quality

Quality of components should affect success rate in crafting as well as unique extra attributes in the end result.

_Note: Components should be able to be combined to form a component of a higher rarity at a `10:1` ratio._

* Junk - The worst result. Cannot be used for crafting, cannot be combined to form higher quality components.
* Rough
* Common
* Superior
* Flawless

So in this way, it would take `10x10x10=1000` Rough components to trade-up to a Flawless component.

If all component permutations are possessed it would result in 12 distinct item types in their inventory:

```
Rough Organic Component (77)
Common Organic Component (24)
Superior Organic Component (6)
Flawless Organic Component (1)

Rough Inorganic Component (77)
Common Inorganic Component (24)
Superior Inorganic Component (6)
Flawless Inorganic Component (1)

Rough Magical Component (77)
Common Magical Component (24)
Superior Magical Component (6)
Flawless Magical Component (1)
```

This may be a bit much, and it may be preferable to track these as a statistic that is not an inventory item. This would introduce complexity in sharing/selling/dropping components for other players, but not insurmountable.

**Comment by ArmoredArtist** [comment]
One thing that comes to mind here is carrying or storing a bunch of ingredients is not fun even if they can be bundled. It clutters your inventory and storage. So having a drop off point for them at a crafting station would fix that and give a centralized area to craft in. You could check your station to see the resources it has available inside of it (the items you found and added to the station). So naturally you would have crafting commands that work while at the crafting station only for all of your crafting needs. This should also be where you bring items back to break down into components (tools of various quality can also be found, bought or made to add to this station to allow more things to be broken down as well as more efficiently). Once broke down, it auto adds the items to the station. When crafting have the station auto use the best ingredients first if your going to do ingredient qualities (with options to change this in table settings). This would neaten everything up into a single area to do your crafting. So instead of having to go to a specific area to craft certain items it would be one central hub. Maybe this could be a work room in a workman's guild or something to that effect. TLDR - All crafting done in a single place that also stores your crafting materials both tools and components.

**Comment by Volte6** [comment]
Yeah I think that's a big takeaway from this... one of each type even if stacked is 12 objects.

The only time this really matters is if you want to be able to give away or share them.

I think being able to craft an object that can be cashed (or salvaged) for them as needed that would work. It could be shared with people that way.

**Reply by ArmoredArtist** [reply]
You could just take the ingredients back out of the work bench (or whatever it is called) as needed to share.

**Reply by Volte6** [reply]
Realistically I'll track it as its own statistic independent of your normal inventory. I can make it droppable (or giveable) and it can spawn an item dynamically. It's more work and a bit more ad-hoc but I don't really mind that too much. I'll just have to figure out where to display that information... making it its own command is easy, making it part of the normal status/score readout is a lot tougher visually. Maybe I can roll it up somehow into their respective categories until you enter a custom command to view them in detail:

`Components: O:13 I:4 M:22`

**Reply by ArmoredArtist** [reply]
You could also have wearable container types that are for ingredients only that give a reason for you to have a bunch of them on you that are not in your direct inventory with a special command to check them.

**Reply by windu-ant** [reply]
I think a reference in score and then a hint at another command to see would be best. Like you said with the O:1 etc. This would make them obvious and keep them visible at a glance but you could also see a more detailed status screen elsewhere. 

O:1 M:2
type 'materials' for a full list
For example

If tools become involved. Blacksmiths hammer, pickaxe, etc, I think some sort of tool belt inventory would make life a lot easier instead of swapping what you're wielding around, when the game could handle this when you use an associated command. "You pull your pickaxe from your tool belt and begin to mine the ore."

**Comment by windu-ant** [comment]
"Crafting should be a time consuming process. While it's not fun to require a player to sit in a room for hours waiting for a result, it should be fine to require them to return at a later date to see the result."

Idea here for creating a time sink would be to cause certain parts of the process to be hands off and time based while crafting the item itself would be much shorter but require the player to be present. Tanning a hide or smelting ore could be the lengthy part and then forging a sword or crafting a vest could be the quicker part once you have the materials collected at a workbench.

**Comment by Volte6** [comment]
I think this negates the time consuming part - effectively collecting the components and moving them into storage AKA smelting would become the "time consuming" part and a single action.

Then it's instantaneous "make whatever you want with everything you've collected", which feels kinda funky to me for some reason.

However, the wait time doesn't have to be a requirement, I just think the skill becomes more interesting and special if you have to wait for results.
